File: C++/프로그래머스/2/49993. 스킬트리/스킬트리.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

int solution(string skill, vector<string> skill_trees) {
    unordered_map<char, int> skill_map;
    for(int i = 0; i < skill.size(); i++) {
        skill_map.insert({skill[i], i});
    }
    
    int answer = 0, sequence = -1;
    bool flag;
    
    for(auto& skill_tree : skill_trees) {
        flag = true; sequence = -1;
        for(auto& st : skill_tree) {
            if(skill_map.find(st) != skill_map.end()) { //스킬트리에 해당 스킬 존재 시
                if(skill_map[st] != ++sequence) {
                    flag = false;
                    break;
                }
            }
        }
        answer += int(flag);
    }
    
    return answer;
}

File: Python3/프로그래머스/2/49993. 스킬트리/스킬트리.py
def solution(skill, skill_trees):
    skill_dict = {skill[i] : i for i in range(len(skill))}
    print(skill_dict)
    answer = 0
    
    sequence, flag = -1, True
    for skill_tree in skill_trees:
        sequence, flag = -1, True
        for st in skill_tree:
            if st in skill_dict:
                if skill_dict[st] != sequence + 1:
                    flag = False
                    break
                sequence = skill_dict[st]
        # print("possible" if flag else "impossible")
        answer += int(flag)
    return answer